### ISSUE TITLE
Crash on exit

### ISSUE THREAD

**Issue by d4k0** [canton7/SyncTrayzor#15]
SyncTrayzor crashes very rarely when exiting the program ("The program encountered a problem and needs to close"). This is the error message that is written to debug.log (I already have several of them there, whereby the first two numbers are changing):

````
[0308/221734:ERROR:renderer_main.cc(227)] Running without renderer sandbox
````

It seems that this error message is related to the WebKit component that is used to display the Syncthing GUI. This also seems to affect "popular" software like Steam, but it seems that there isn't any "solution" at the moment (the problem seems to occur when the WebKit component is used in a certain way). Unfortunately, this crash seems to occur randomly, so I haven't found any steps yet to reproduce it (Synthing is however stopped properly before the crash occurs).

**Comment by canton7** [comment]
This is when closing Syncthing, as opposed to logging off / shutting down windows? (There's a similar issue where the embedded Chromium component crashes hard in this scenario, see cefsharp/CefSharp#800).

Is there more to the debug log that you can paste (giving me more googlable terms)? Is there anything in `%APPDATA\SyncTrayzor\SyncTrayzor.log` that's relevant?

**Comment by canton7** [comment]
Oh, and anything in the event log? Crashes normally end up with something in there....

**Comment by d4k0** [comment]
Well, I got this crash only twice so far, but if I remember correctly, this always happened when SyncTrayzor was minimized to the system tray (no sync was occuring, Syncthing was idling) and I stopped it by right-clicking on the icon and choosing "Exit". This is the whole debug.log:

```
[0307/215321:ERROR:renderer_main.cc(227)] Running without renderer sandbox
[0307/215436:ERROR:renderer_main.cc(227)] Running without renderer sandbox
[0307/215616:ERROR:renderer_main.cc(227)] Running without renderer sandbox
[0307/215905:ERROR:renderer_main.cc(227)] Running without renderer sandbox
[0307/220029:ERROR:renderer_main.cc(227)] Running without renderer sandbox
[0307/220455:ERROR:renderer_main.cc(227)] Running without renderer sandbox
[0308/162633:ERROR:renderer_main.cc(227)] Running without renderer sandbox
[0308/163002:ERROR:renderer_main.cc(227)] Running without renderer sandbox
[0308/221218:ERROR:renderer_main.cc(227)] Running without renderer sandbox
[0308/221734:ERROR:renderer_main.cc(227)] Running without renderer sandbox
```

So nothing new here. SyncTrayzor.log also doesn't show anything interesting as the last entry there is written a few minutes before the crash.

However, I totally forgot about the event log (thanks for mentioning it!) which in fact shows two errors when the crash happened:

**Red error with ".NET Runtime" as source:**
```
Anwendung: SyncTrayzor.exe
Frameworkversion: v4.0.30319
Beschreibung: Der Prozess wurde aufgrund eines Ausnahmefehlers beendet.
Ausnahmeinformationen: System.Reflection.TargetInvocationException
Stapel:
   bei System.RuntimeMethodHandle.InvokeMethod(System.Object, System.Object[], System.Signature, Boolean)
   bei System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(System.Object, System.Object[], System.Object[])
   bei System.Reflection.RuntimeMethodInfo.Invoke(System.Object, System.Reflection.BindingFlags, System.Reflection.Binder, System.Object[], System.Globalization.CultureInfo)
   bei Stylet.ViewManager.CreateViewForModel(System.Object)
   bei Stylet.ViewManager.CreateAndBindViewForModel(System.Object)
   bei Stylet.WindowManager.CreateWindow(System.Object, Boolean)
   bei Stylet.WindowManager.ShowDialog(System.Object)
   bei System.Windows.Threading.Dispatcher.CatchException(System.Exception)
   bei MS.Internal.Threading.ExceptionFilterHelper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
   bei System.Windows.Threading.Dispatcher.LegacyInvokeImpl(System.Windows.Threading.DispatcherPriority, System.TimeSpan, System.Delegate, System.Object, Int32)
   bei MS.Win32.HwndSubclass.SubclassWndProc(IntPtr, Int32, IntPtr, IntPtr)
   bei MS.Win32.UnsafeNativeMethods.DispatchMessage(System.Windows.Interop.MSG ByRef)
   bei MS.Win32.UnsafeNativeMethods.DispatchMessage(System.Windows.Interop.MSG ByRef)
   bei System.Windows.Threading.Dispatcher.PushFrameImpl(System.Windows.Threading.DispatcherFrame)
   bei System.Windows.Application.RunInternal(System.Windows.Window)
   bei System.Windows.Application.Run()
   bei SyncTrayzor.App.Main()
```
**Red error with "Application Error" as source:**
```
Name der fehlerhaften Anwendung: SyncTrayzor.exe, Version: 1.0.6.0, Zeitstempel: 0x54fbc1d8
Name des fehlerhaften Moduls: KERNELBASE.dll, Version: 6.3.9600.17415, Zeitstempel: 0x54505737
Ausnahmecode: 0xe0434352
Fehleroffset: 0x0000000000008b9c
ID des fehlerhaften Prozesses: 0x2320
Startzeit der fehlerhaften Anwendung: 0x01d059e54a74ba8d
Pfad der fehlerhaften Anwendung: D:\Programme\Internet\SyncTrayzorPortable\SyncTrayzor.exe
Pfad des fehlerhaften Moduls: C:\Windows\system32\KERNELBASE.dll
Berichtskennung: f50023a6-c5d8-11e4-8380-00241dce9ad3
Vollständiger Name des fehlerhaften Pakets: 
Anwendungs-ID, die relativ zum fehlerhaften Paket ist:
```

Here is some discussion about the error:
https://code.google.com/p/chromiumembedded/issues/detail?id=524#c4
https://github.com/atom/atom-shell/issues/22
https://github.com/ValveSoftware/steam-for-linux/issues/3620

However, after reading the second link, I'm not so sure anymore that this crash was really caused by the WebKit component :-/. On the other hand, "MS.Win32.UnsafeNativeMethods.DispatchMessage" for example seems to point to WPF (or rendering a Windows in general).

**Comment by d4k0** [comment]
I also found the second crash in the event log (this was for version 1.0.5 of Synctrayzor):

**Red error with ".NET Runtime" as source:**
````
Anwendung: SyncTrayzor.exe
Frameworkversion: v4.0.30319
Beschreibung: Der Prozess wurde aufgrund eines Ausnahmefehlers beendet.
Ausnahmeinformationen: System.ObjectDisposedException
Stapel:
   bei StyletIoC.Internal.Container.GetRegistrations(StyletIoC.Internal.TypeKey, Boolean)
   bei StyletIoC.Internal.Container.Get[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]](System.String)
   bei SyncTrayzor.Bootstrapper.OnUnhandledException(System.Windows.Threading.DispatcherUnhandledExceptionEventArgs)
   bei System.Windows.Threading.Dispatcher.CatchException(System.Exception)
   bei MS.Internal.Threading.ExceptionFilterHelper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
   bei System.Windows.Threading.Dispatcher.LegacyInvokeImpl(System.Windows.Threading.DispatcherPriority, System.TimeSpan, System.Delegate, System.Object, Int32)
   bei MS.Win32.HwndSubclass.SubclassWndProc(IntPtr, Int32, IntPtr, IntPtr)
   bei MS.Win32.UnsafeNativeMethods.DispatchMessage(System.Windows.Interop.MSG ByRef)
   bei MS.Win32.UnsafeNativeMethods.DispatchMessage(System.Windows.Interop.MSG ByRef)
   bei System.Windows.Threading.Dispatcher.PushFrameImpl(System.Windows.Threading.DispatcherFrame)
   bei System.Windows.Application.RunInternal(System.Windows.Window)
   bei System.Windows.Application.Run()
   bei SyncTrayzor.App.Main()
````

**Red error with "Application Error" as source:**
````
Name der fehlerhaften Anwendung: SyncTrayzor.exe, Version: 1.0.5.0, Zeitstempel: 0x54fb22ec
Name des fehlerhaften Moduls: KERNELBASE.dll, Version: 6.3.9600.17415, Zeitstempel: 0x54505737
Ausnahmecode: 0xe0434352
Fehleroffset: 0x0000000000008b9c
ID des fehlerhaften Prozesses: 0x1830
Startzeit der fehlerhaften Anwendung: 0x01d05918bd8c7882
Pfad der fehlerhaften Anwendung: D:\Programme\Internet\SyncTrayzorPortable\SyncTrayzor.exe
Pfad des fehlerhaften Moduls: C:\Windows\system32\KERNELBASE.dll
Berichtskennung: 0d3af21d-c50c-11e4-8380-00241dce9ad3
Vollständiger Name des fehlerhaften Pakets: 
Anwendungs-ID, die relativ zum fehlerhaften Paket ist:
````

This crash seems to be linked to the IoC container you are using, but it also for example contains the "MS.Win32.UnsafeNativeMethods.DispatchMessage" entry in the trace stack.

**Comment by canton7** [comment]
Thanks! That 3rd stack trace, with the IoC container in it, is because we get an exception on shutdown. The IoC container's already been disposed, but then we try and create a dialog to display an exception... And get another exception. We should have logged to SyncTrayzor.log to that point, but maybe it hadn't flushed - I'll fix that. This means that the exception is occurring after the application's started exiting.

The other .NET exception - the 1st stack trace - is similar. We're trying to display the 'unhandled exception' dialog, and falling over (since this is on application shutdown, this is not so surprising). But again we should have logged the exception by this point. 

I'm not sure whether the "Application Error" stack traces are the root cause of the problem, or are as a result of the unhandled exception handler throwing an exception. I'll:

1. Flush the log after writing an error. That should give us a decent "original error" stack trace.
2. Not try and display an error dialog if the application is shutting down. This should prevent that second crash.

**Comment by d4k0** [comment]
Regarding the 3rd stack trace: I think I could reproduce this crash. It happens when one exits SyncTrazor before everything is loaded completely (especially the GUI of Syncthing).

So far I couldn't reproduce the first crash (with the first stack trace), but as you said it should be something similiar.

Regarding the "Application Error" stack traces: I think they are a result of the unhandled exception as the event log says each "Application Error" and its appropriate .NET Runtime error occured at the exact same time.

All in all I'm not sure anymore that these crashes have anything to do with the WebKit component because the error message I mentioned in my first post always seems to be written to the debug.log when there's a crash (maybe I should rename the title?).

**Comment by canton7** [comment]
Re the "Application Error" - there's two exceptions going on here. There's a first unhandled exception. SyncTrayzor tries to handle this, in order to log it and display a nice dialog to the user. This second logging-and-displaying step then fails with an error (because the application is shutting down).

The .NET stack trace you see is the second exception - the one we encounter while trying to display the first exception. I'm not sure whether the Application Error refers to the first exception or the second one.

I tried following your instructions to reproduce without success. If it's working for you, however, then great. Wait for the next release (hopefully coming out tomorrow), repro, and let's see what the error log contains that time.

**Comment by canton7** [comment]
1.0.7 contains the logging we need. Would you mind upgrading, and report back if/when you mange to repro/see it happen again? Thanks!

**Comment by d4k0** [comment]
Ok, here are the interesting parts from Synctrazor.log (from the "logs" subfolder):

````
2015-03-10 15:49:33.4690 [Debug] SyncTrayzor.SyncThing.SyncThingProcessRunner: [JEMTR] 15:49:33 INFO: Starting web GUI on http://127.0.0.1:8384/ 
2015-03-10 15:49:33.7112 [Info] SyncTrayzor.SyncThing.SyncThingApiClient: Requesting API shutdown 
2015-03-10 15:49:33.9720 [Debug] SyncTrayzor.SyncThing.SyncThingProcessRunner: [JEMTR] 15:49:33 INFO: Starting UPnP discovery... 
2015-03-10 15:49:34.2239 [Info] SyncTrayzor.SyncThing.SyncThingProcessRunner: Syncthing process stopped with exit status -1 
2015-03-10 15:49:34.2239 [Error] SyncTrayzor.Bootstrapper: An unhandled exception occurred System.ComponentModel.Win32Exception Zugriff verweigert Void Kill()    bei System.Diagnostics.Process.Kill()
   bei StyletIoC.Internal.Registrations.SingletonRegistration.<.ctor>b__0(Object o, EventArgs e)
   bei System.EventHandler.Invoke(Object sender, EventArgs e)
   bei Stylet.Bootstrapper`1.Dispose()
   bei System.Windows.ExitEventHandler.Invoke(Object sender, ExitEventArgs e)
   bei System.Windows.Application.DoShutdown()
   bei System.Windows.Application.ShutdownImpl()
   bei System.Windows.Application.ShutdownCallback(Object arg)
   bei System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   bei MS.Internal.Threading.ExceptionFilterHelper.TryCatchWhen(Object source, Delegate method, Object args, Int32 numArgs, Delegate catchHandler) System.ComponentModel.Win32Exception (0x80004005): Zugriff verweigert
   bei System.Diagnostics.Process.Kill()
   bei StyletIoC.Internal.Registrations.SingletonRegistration.<.ctor>b__0(Object o, EventArgs e)
   bei System.EventHandler.Invoke(Object sender, EventArgs e)
   bei Stylet.Bootstrapper`1.Dispose()
   bei System.Windows.ExitEventHandler.Invoke(Object sender, ExitEventArgs e)
   bei System.Windows.Application.DoShutdown()
   bei System.Windows.Application.ShutdownImpl()
   bei System.Windows.Application.ShutdownCallback(Object arg)
   bei System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   bei MS.Internal.Threading.ExceptionFilterHelper.TryCatchWhen(Object source, Delegate method, Object args, Int32 numArgs, Delegate catchHandler)
````
It seems that stopping/killing the Syncthing process leads to this exception. As far as I can tell the IoC container is also involved here, so maybe the singleton (which I guess is a reference to the Syncthing process) isn't initalized yet (because I stopped Synctrayzor while Syncthing was still starting)?

But as I said this crash only occurs when I exit Synctrayzor as fast as possible after the start of the program, so this maybe doesn't reflect the "normal" use of the program :).

**Comment by canton7** [comment]
Still worth fixing! The IoC container is only there before disposing it disposes all the singletons it owns, and one of those singletons's Dispose methods is throwing an exception. 

What happens is that the Syncthing process dies (because we ask it to shut down). We normally detect the fact that it's dead and clear our reference to its process, but for some reason that isn't happening here (there's no lock on that field, so "threading" is probably to blame). We then try and kill it *again*, and that's falling over with a Win32Exception with 'Access Denied', which the docs say can occur if "The process is terminating." - which looks like the case here.

So, it's harmless, but we shouldn't be crashing! I'll put a proper memory barrier around that field, and catch Win32Exceptions there just to be sure.

**Comment by canton7** [comment]
Right, that commit should fix it, when it makes it into a release.

When that release comes out, even if SyncTrayzor doesn't crash, I'd still be interested in seeing the log after you've tried to repro. I want to see whether the "Killinternal failed with an error" message appears.

**Comment by d4k0** [comment]
OK, I'll try it again when the new version is released.

If there are still any crashes (and I'm the only one to reproduce them), I'll probably try to debug them as I have installed Visual Studio 2013 anyway.

**Comment by canton7** [comment]
Great, thank you!

**Comment by canton7** [comment]
Ack, bad choice of words in the commit comment - didn't mean to close the issue.

Think I tracked it down. Cause was closing SyncTrayzor after never having run Syncthing. Should be fixed in v1.0.8, just released.

**Comment by d4k0** [comment]
I can confirm that this crash seems to be fixed. Thanks!

I haven't been able to reproduce the first crash ("System.Reflection.TargetInvocationException"), so I'm not sure if this was also fixed (as a side effect) or if this is another issue. I'll propably open a new issue if I encounter the first crash again (and can reproduce it), so I think this issue can be closed.

**Comment by canton7** [comment]
There were two issues here. The first, root, issue is the crash on exit, which a5055b0 fixed. Then there was a second issue where we tried to handle this error (displaying a dialog to the user), but *that* crashed because we were shutting down when we tried to do that. The first few stack traces we looked at were this second "derivative" error - different things were failing, but all because the application was exiting at the time. This second "derivative" error was masking the first - its stack trace was being recorded, instead of the first error's stack trace - so I had to fix that first.
